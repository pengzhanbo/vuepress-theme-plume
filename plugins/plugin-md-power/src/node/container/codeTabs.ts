import type { PluginWithOptions } from 'markdown-it'
import type { CodeTabsOptions } from '../../shared/index.js'
import { tab } from '@mdit/plugin-tab'
import { isPlainObject } from '@vuepress/helper'
import { definitions, getFileIconName, getFileIconTypeFromExtension } from '../fileIcons/index.js'
import { stringifyProp } from '../utils/stringifyProp.js'

export const codeTabs: PluginWithOptions<CodeTabsOptions> = (md, options: CodeTabsOptions = {}) => {
  const getIcon = (filename: string): string | void => {
    if (options.icon === false)
      return undefined
    const { named, extensions } = isPlainObject(options.icon) ? options.icon : {}
    if (named === false && definitions.named[filename])
      return undefined
    if (extensions === false && getFileIconTypeFromExtension(filename)) {
      return undefined
    }
    const hasNamed = named && named.length
    const hasExt = extensions && extensions.length
    if (hasNamed || hasExt) {
      if (hasNamed && named.includes(filename))
        return definitions.named[filename]
      if (hasExt && extensions.some(ext => filename.endsWith(ext)))
        return getFileIconTypeFromExtension(filename)
      return undefined
    }
    return getFileIconName(filename)
  }

  tab(md, {
    name: 'code-tabs',

    tabsOpenRenderer: ({ active, data }, tokens, index) => {
      const { meta } = tokens[index]
      const titles = data.map(({ title }) => md.renderInline(title))
      const tabsData = data.map((item, dataIndex) => {
        const { id = titles[dataIndex] } = item

        return { id }
      })

      const titlesContent = titles.map((title, index) => {
        const icon = getIcon(title)
        return `<template #title${index}="{ value, isActive }">${icon ? `<VPIcon name="${icon}"/>` : ''}<span>${title}</span></template>`
      }).join('')

      return `<CodeTabs id="${index}" :data='${stringifyProp(tabsData)}'${active === -1 ? '' : ` :active="${active}"`}${meta.id ? ` tab-id="${meta.id as string}"` : ''}>${titlesContent}`
    },

    tabsCloseRenderer: () => `</CodeTabs>`,

    tabOpenRenderer: ({ index }, tokens, tokenIndex) => {
      let foundFence = false

      // Hide all elements excerpt the first fence
      for (let i = tokenIndex; i < tokens.length; i++) {
        const { type } = tokens[i]

        if (type === 'code-tabs_tab_close')
          break

        if ((type === 'fence' || type === 'import_code') && !foundFence) {
          foundFence = true
          continue
        }

        tokens[i].type = 'code_tab_empty'
        tokens[i].hidden = true
      }

      return `<template #tab${index}="{ value, isActive }">`
    },

    tabCloseRenderer: () => `</template>`,
  })
}
