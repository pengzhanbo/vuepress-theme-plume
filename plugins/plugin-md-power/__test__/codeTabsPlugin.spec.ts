import type { CodeTabsOptions } from '../src/shared/codeTabs.js'
import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { codeTabs } from '../src/node/container/codeTabs.js'

function createMarkdown(options?: CodeTabsOptions) {
  const md = new MarkdownIt()
  md.options.highlight = str => `<pre><code>${str}</code></pre>`
  const fence = md.renderer.rules.fence!
  md.renderer.rules.fence = (...args) => `<div class="language-lang">${fence(...args)}</div>`

  md.use(codeTabs, options)
  return md
}

const FENCE = '```'

describe('codeTabsPlugin', () => {
  const code = `\
::: code-tabs
@tab tab-1
${FENCE}js
const a = 1
${FENCE}
@tab tab-2
${FENCE}js
const b = 2
${FENCE}
@tab tab-3
${FENCE}js
const c = 3
${FENCE}
:::

::: code-tabs
@tab npm
${FENCE}sh
npm i
${FENCE}
@tab:active pnpm
${FENCE}sh
pnpm i
${FENCE}
@tab yarn
${FENCE}sh
yarn
${FENCE}
:::

::: code-tabs#pm
@tab a.ts
${FENCE}ts
const a = 1
${FENCE}
@tab a.js
${FENCE}js
const a = 1
${FENCE}
:::
`
  it('should work with default', () => {
    const md = createMarkdown()
    expect(md.render(code)).toMatchSnapshot()
  })

  it('should work with no icon', () => {
    const md = createMarkdown({ icon: false })
    expect(md.render(code)).toMatchSnapshot()
  })

  it('should work with options: `{ named: false, extensions: false }`', () => {
    const md = createMarkdown({ icon: { named: false, extensions: false } })
    expect(md.render(code)).toMatchSnapshot()
  })

  it('should work with options: { named: [npm,pnpm,yarn], extensions: [.js,.ts] }', () => {
    const md = createMarkdown({ icon: { named: ['npm', 'pnpm', 'yarn'], extensions: ['.js', '.ts'] } })
    expect(md.render(code)).toMatchSnapshot()
  })
})
