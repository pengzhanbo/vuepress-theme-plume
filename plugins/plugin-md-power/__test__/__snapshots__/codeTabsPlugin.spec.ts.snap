// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codeTabsPlugin > should work with default 1`] = `
"<CodeTabs id="0" :data='[{"id":"tab-1"},{"id":"tab-2"},{"id":"tab-3"}]'><template #title0="{ value, isActive }"><span>tab-1</span></template><template #title1="{ value, isActive }"><span>tab-2</span></template><template #title2="{ value, isActive }"><span>tab-3</span></template><template #tab0="{ value, isActive }"><div class="language-lang"><pre><code>const a = 1
</code></pre>
</div></template><template #tab1="{ value, isActive }"><div class="language-lang"><pre><code>const b = 2
</code></pre>
</div></template><template #tab2="{ value, isActive }"><div class="language-lang"><pre><code>const c = 3
</code></pre>
</div></template></CodeTabs><CodeTabs id="11" :data='[{"id":"npm"},{"id":"pnpm"},{"id":"yarn"}]' :active="1"><template #title0="{ value, isActive }"><VPIcon name="logos:npm-icon"/><span>npm</span></template><template #title1="{ value, isActive }"><VPIcon name="vscode-icons:file-type-light-pnpm"/><span>pnpm</span></template><template #title2="{ value, isActive }"><VPIcon name="vscode-icons:file-type-yarn"/><span>yarn</span></template><template #tab0="{ value, isActive }"><div class="language-lang"><pre><code>npm i
</code></pre>
</div></template><template #tab1="{ value, isActive }"><div class="language-lang"><pre><code>pnpm i
</code></pre>
</div></template><template #tab2="{ value, isActive }"><div class="language-lang"><pre><code>yarn
</code></pre>
</div></template></CodeTabs><CodeTabs id="22" :data='[{"id":"a.ts"},{"id":"a.js"}]' tab-id="pm"><template #title0="{ value, isActive }"><VPIcon name="vscode-icons:file-type-typescript"/><span>a.ts</span></template><template #title1="{ value, isActive }"><VPIcon name="vscode-icons:file-type-js"/><span>a.js</span></template><template #tab0="{ value, isActive }"><div class="language-lang"><pre><code>const a = 1
</code></pre>
</div></template><template #tab1="{ value, isActive }"><div class="language-lang"><pre><code>const a = 1
</code></pre>
</div></template></CodeTabs>"
`;

exports[`codeTabsPlugin > should work with no icon 1`] = `
"<CodeTabs id="0" :data='[{"id":"tab-1"},{"id":"tab-2"},{"id":"tab-3"}]'><template #title0="{ value, isActive }"><span>tab-1</span></template><template #title1="{ value, isActive }"><span>tab-2</span></template><template #title2="{ value, isActive }"><span>tab-3</span></template><template #tab0="{ value, isActive }"><div class="language-lang"><pre><code>const a = 1
</code></pre>
</div></template><template #tab1="{ value, isActive }"><div class="language-lang"><pre><code>const b = 2
</code></pre>
</div></template><template #tab2="{ value, isActive }"><div class="language-lang"><pre><code>const c = 3
</code></pre>
</div></template></CodeTabs><CodeTabs id="11" :data='[{"id":"npm"},{"id":"pnpm"},{"id":"yarn"}]' :active="1"><template #title0="{ value, isActive }"><span>npm</span></template><template #title1="{ value, isActive }"><span>pnpm</span></template><template #title2="{ value, isActive }"><span>yarn</span></template><template #tab0="{ value, isActive }"><div class="language-lang"><pre><code>npm i
</code></pre>
</div></template><template #tab1="{ value, isActive }"><div class="language-lang"><pre><code>pnpm i
</code></pre>
</div></template><template #tab2="{ value, isActive }"><div class="language-lang"><pre><code>yarn
</code></pre>
</div></template></CodeTabs><CodeTabs id="22" :data='[{"id":"a.ts"},{"id":"a.js"}]' tab-id="pm"><template #title0="{ value, isActive }"><span>a.ts</span></template><template #title1="{ value, isActive }"><span>a.js</span></template><template #tab0="{ value, isActive }"><div class="language-lang"><pre><code>const a = 1
</code></pre>
</div></template><template #tab1="{ value, isActive }"><div class="language-lang"><pre><code>const a = 1
</code></pre>
</div></template></CodeTabs>"
`;

exports[`codeTabsPlugin > should work with options: \`{ named: false, extensions: false }\` 1`] = `
"<CodeTabs id="0" :data='[{"id":"tab-1"},{"id":"tab-2"},{"id":"tab-3"}]'><template #title0="{ value, isActive }"><span>tab-1</span></template><template #title1="{ value, isActive }"><span>tab-2</span></template><template #title2="{ value, isActive }"><span>tab-3</span></template><template #tab0="{ value, isActive }"><div class="language-lang"><pre><code>const a = 1
</code></pre>
</div></template><template #tab1="{ value, isActive }"><div class="language-lang"><pre><code>const b = 2
</code></pre>
</div></template><template #tab2="{ value, isActive }"><div class="language-lang"><pre><code>const c = 3
</code></pre>
</div></template></CodeTabs><CodeTabs id="11" :data='[{"id":"npm"},{"id":"pnpm"},{"id":"yarn"}]' :active="1"><template #title0="{ value, isActive }"><span>npm</span></template><template #title1="{ value, isActive }"><span>pnpm</span></template><template #title2="{ value, isActive }"><span>yarn</span></template><template #tab0="{ value, isActive }"><div class="language-lang"><pre><code>npm i
</code></pre>
</div></template><template #tab1="{ value, isActive }"><div class="language-lang"><pre><code>pnpm i
</code></pre>
</div></template><template #tab2="{ value, isActive }"><div class="language-lang"><pre><code>yarn
</code></pre>
</div></template></CodeTabs><CodeTabs id="22" :data='[{"id":"a.ts"},{"id":"a.js"}]' tab-id="pm"><template #title0="{ value, isActive }"><span>a.ts</span></template><template #title1="{ value, isActive }"><span>a.js</span></template><template #tab0="{ value, isActive }"><div class="language-lang"><pre><code>const a = 1
</code></pre>
</div></template><template #tab1="{ value, isActive }"><div class="language-lang"><pre><code>const a = 1
</code></pre>
</div></template></CodeTabs>"
`;

exports[`codeTabsPlugin > should work with options: { named: [npm,pnpm,yarn], extensions: [.js,.ts] } 1`] = `
"<CodeTabs id="0" :data='[{"id":"tab-1"},{"id":"tab-2"},{"id":"tab-3"}]'><template #title0="{ value, isActive }"><span>tab-1</span></template><template #title1="{ value, isActive }"><span>tab-2</span></template><template #title2="{ value, isActive }"><span>tab-3</span></template><template #tab0="{ value, isActive }"><div class="language-lang"><pre><code>const a = 1
</code></pre>
</div></template><template #tab1="{ value, isActive }"><div class="language-lang"><pre><code>const b = 2
</code></pre>
</div></template><template #tab2="{ value, isActive }"><div class="language-lang"><pre><code>const c = 3
</code></pre>
</div></template></CodeTabs><CodeTabs id="11" :data='[{"id":"npm"},{"id":"pnpm"},{"id":"yarn"}]' :active="1"><template #title0="{ value, isActive }"><VPIcon name="logos:npm-icon"/><span>npm</span></template><template #title1="{ value, isActive }"><VPIcon name="vscode-icons:file-type-light-pnpm"/><span>pnpm</span></template><template #title2="{ value, isActive }"><VPIcon name="vscode-icons:file-type-yarn"/><span>yarn</span></template><template #tab0="{ value, isActive }"><div class="language-lang"><pre><code>npm i
</code></pre>
</div></template><template #tab1="{ value, isActive }"><div class="language-lang"><pre><code>pnpm i
</code></pre>
</div></template><template #tab2="{ value, isActive }"><div class="language-lang"><pre><code>yarn
</code></pre>
</div></template></CodeTabs><CodeTabs id="22" :data='[{"id":"a.ts"},{"id":"a.js"}]' tab-id="pm"><template #title0="{ value, isActive }"><VPIcon name="vscode-icons:file-type-typescript"/><span>a.ts</span></template><template #title1="{ value, isActive }"><VPIcon name="vscode-icons:file-type-js"/><span>a.js</span></template><template #tab0="{ value, isActive }"><div class="language-lang"><pre><code>const a = 1
</code></pre>
</div></template><template #tab1="{ value, isActive }"><div class="language-lang"><pre><code>const a = 1
</code></pre>
</div></template></CodeTabs>"
`;
